modules/cudaarithm/src/cuda/transpose.cu

```diff
@@ -60,34 +60,147 @@ void cv::cuda::transpose(InputArray _src, OutputArray _dst, Stream& stream)
 {
     GpuMat src = getInputMat(_src, stream);
 
+    const int srcType = src.type();
+    const int srcDepth = src.depth();
+    const int srcCn = src.channels();
     const size_t elemSize = src.elemSize();
+    const size_t elemSize1 = src.elemSize1();
 
-    CV_Assert( elemSize == 1 || elemSize == 4 || elemSize == 8 );
+    //CV_Assert( elemSize == 1 || elemSize == 4 || elemSize == 8 );
 
     GpuMat dst = getOutputMat(_dst, src.cols, src.rows, src.type(), stream);
 
-    if (elemSize == 1)
+    const bool isNppiNativelySupported =
+      (srcType == CV_8UC1)  || (srcType == CV_8UC3)  || (srcType == CV_8UC4)  ||
+      (srcType == CV_16UC1) || (srcType == CV_16UC3) || (srcType == CV_16UC4) ||
+      (srcType == CV_16SC1) || (srcType == CV_16SC3) || (srcType == CV_16SC4) ||
+      (srcType == CV_32SC1) || (srcType == CV_32SC3) || (srcType == CV_32SC4) ||
+      (srcType == CV_32FC1) || (srcType == CV_32FC3) || (srcType == CV_32FC4);
+    const bool isElemSizeSupportedByNppi =
+      ((elemSize != 0) && !(elemSize%1) && ((elemSize/1)<=4)) ||
```

Comment:
Isn't elemSize%1 == 0 always?
When can elemSize == 0?

Reply:
You are right, but it was on purpose for code clarity/readibility
elemSize%1 and elemSize/1 will be optimized out by the compiler, this is just to balance with the 2, 4, 8 cases

Ok to get rid of the (elemSize != 0) check

Reply:
For clarity I would use elemSize1 and channels?
Under what circumstance will `isNppiNativelySupported  == false` and `isElemSizeSupportedByNppi == true` or `isElemSizeSupportedByGridTranspose  == true`?

Reply:
There is overlap between isNppiNativelySupported, isElemSizeSupportedByNppi  and isElemSizeSupportedByGridTranspose 
The idea is that if anything goes wrong, they can be disabled independently.

Reply:
Is there a complete overlap, i.e. is `isElemSizeSupportedByNppi` and `isElemSizeSupportedByGridTranspose` redundant unless
`isNppiNativelySupported` is manually set to `false`?

Reply:
> Is there a complete overlap, i.e. is `isElemSizeSupportedByNppi` and `isElemSizeSupportedByGridTranspose` redundant unless
> `isNppiNativelySupported` is manually set to `false`?

There are cases where (isNppiNativelySupported == false) and (isElemSizeSupportedByNppi== true) (e.g. 16U_C2)

But there are currently no cases where (isElemSizeSupportedByNppi== false) and  (isElemSizeSupportedByGridTranspose == true). This could be useful if performance tests reveals that some NPP calls are slower than gridTranspose(). Specific cases could be excluded from isElemSizeSupportedByNppi.

Reply:
Is `16UC2` is handled by the below block?
```
else if (!(elemSize%2) && ((elemSize/2)==2))
  nppSafeCall( nppiTranspose_8u_C4R(src.ptr<Npp8u>(), static_cast<int>(src.step),
    dst.ptr<Npp8u>(), static_cast<int>(dst.step), sz) );
```
If so wouldn't it be better to have this under 
```
else if (srcType == CV_16UC2)
```

Reply:
Exactly, this is not a native nppi Call (isNppiNativelySupported == false) but it can be handled by cheatin on the data type (isElemSizeSupportedByNppi  == true)
That's why the tests inside `if (isElemSizeSupportedByNppi)` is explicitely focusing on `elemSize `rather than `srcType`

Reply:
I'm not sure if this makes it easier or harder for the next person who looks at it to see whats going on, what do you think @alalek ?

Either way I think having overlap between both makes it difficult to understand, maybe if the "cheat" only included the cases where it was applicable it would be more obvious?